kata_proxy.go

```diff
@@ -0,0 +1,98 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"os/exec"
+	"path/filepath"
+)
+
+// This is the Kata Containers implementation of the proxy interface.
+// This is pretty simple since it provides the same interface to both
+// runtime and shim as if they were talking directly to the agent.
+type kataProxy struct {
+	proxyURL string
+}
+
+// start is kataProxy start implementation for proxy interface.
+func (p *kataProxy) start(pod Pod) (int, string, error) {
+	config, err := newProxyConfig(pod.config)
+	if err != nil {
+		return -1, "", err
+	}
+
+	// construct the socket path the proxy instance will use
+	socketPath := filepath.Join(runStoragePath, pod.id, "kata_proxy.sock")
+	proxyURL := fmt.Sprintf("unix://%s", socketPath)
+
+	vmURL, err := pod.agent.vmURL()
```

Comment:
So here we assume `agent.init()` has been called already, right ?

Reply:
Yes. `proxy.start()` is called here: https://github.com/containers/virtcontainers/blob/master/pod.go#L768 from a higher level call here: https://github.com/containers/virtcontainers/blob/master/api.go#L90
And `agent.init()` is called here: https://github.com/containers/virtcontainers/blob/master/pod.go#L653 from https://github.com/containers/virtcontainers/blob/master/pod.go#L561 which is called from here: https://github.com/containers/virtcontainers/blob/master/api.go#L46

Now the question could be, do we want a proper check to make sure the agent has been initialized ? In that case, we might want to introduce `isInitialized()` function to the agent interface.